### PR TITLE
Allow not providing values to selected configurable variables through CLI

### LINKED ISSUE #42859: [Bug]: Can't provide a value to a configurable variable via the CLI when there are other configurable variables for which values cannot be provided via the CLI
### Description

$title.

Also, why can't we provide a value for variables of types like `int?` via the CLI? Here, we surely can support at least int values, even if we don't support nil?

### Steps to Reproduce

```ballerina
import ballerina/io;

configurable int v1 = 1;
configurable int? v2 = ();

public function main() {
    io:println(v1);
    io:println(v2);
}
```

```cmd
$ bal run test.bal -- -Cv1=1 
Compiling source
        test.bal

Running executable

error: value for configurable variable 'v2' with type 'int?' is not supported as a command line argument
        at .:0.0.0(test.bal:4)
```

### Affected Version(s)

2201.9.0

### OS, DB, other environment details and versions

_No response_

### Related area

-> Runtime

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by ravinperera00**
Fixes #42859 

## Approach
**Bug fix**

When considering unsupported configurable values in the CLI, there could be two cases:

Case 1
A type that is not supported through the CLI is used as a configurable variable but the value is not provided through the CLI. This should not be a problem and the CLI should allow leaving the value for that out as long as a default values is provided through other methods.

Case 2
A type that is not supported through the CLI is used as a configurable variable but the value for that is provided through the CLI. This should cause and error stopping the program.

Checking whether the variable value is null before executing the checks regarding the type can produce the expected outcome.

**Improvement**

The Ballerina spec mentions the following regarding the supported types for configurable values.

> Configurable values can be supplied with the built-in command-line option -C,
> 
> `-Ckey=value`
> 
> Key syntax:
> `key:= [[org .] module .] variable`
> 
> Define S as:
> `type S boolean | int | float | decimal | string | xml `
> 
> In the first phase, only the configurable variables with type S? are supported. We can support anydata[], map&lt;anydata> and table&lt;map&lt;anydata>> later as needed.

Therefore types such as int? should be supported through the CLI. This can be achieved by accepting the `()` type as a valid member type within union types.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/ConfigUtils.java [hunk 1]
```diff
@@ -85,7 +85,7 @@ public static boolean containsUnsupportedMembers(BUnionType unionType) {
     }
 
     private static boolean isSimpleSequenceType(int tag) {
-        return tag <= TypeTags.BOOLEAN_TAG || TypeTags.isXMLTypeTag(tag);
+        return (tag == TypeTags.NULL_TAG) || (tag <= TypeTags.BOOLEAN_TAG) || TypeTags.isXMLTypeTag(tag);
```
Comment (HindujaB) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/ConfigUtils.java:88]:
```suggestion
        return tag == TypeTags.NULL_TAG)|| tag <= TypeTags.BOOLEAN_TAG || TypeTags.isXMLTypeTag(tag);
```

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/cli/CliProvider.java [hunk 2]
```diff
@@ -220,12 +220,12 @@ public Optional<ConfigValue> getAsUnionAndMark(Module module, VariableKey key) {
         CliArg cliArg = getCliArg(module, key);
         BUnionType unionType = (BUnionType) ((BIntersectionType) key.type).getEffectiveType();
         boolean isEnum = SymbolFlags.isFlagOn(unionType.getFlags(), SymbolFlags.ENUM);
-        if (!isEnum && ConfigUtils.containsUnsupportedMembers(unionType)) {
-            throw new ConfigException(CONFIG_CLI_TYPE_NOT_SUPPORTED, key.variable, unionType);
-        }
         if (cliArg.value == null) {
             return Optional.empty();
         }
+        if (!isEnum && ConfigUtils.containsUnsupportedMembers(unionType)) {
```
Comment (HindujaB) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/cli/CliProvider.java:226]: Shall we add some tests and modify existing tests accordingly?

**Comment by HindujaB** [comment]
Let's shorten the PR title as well. We can elaborate on it inside the description and approach.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/cli/CliProvider.java (hunk 2 above)
Comment (ravinperera00) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/cli/CliProvider.java:226]: I'll do that.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/ConfigUtils.java (hunk 1 above)
Comment (ravinperera00) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/ConfigUtils.java:88]: I'll refactor this

File: bvm/ballerina-runtime/src/test/java/io/ballerina/runtime/test/config/ConfigTest.java [hunk 3]
```diff
@@ -258,4 +258,25 @@ public void testEnvVarUnsupportedErrors(String variableName, Type type, String e
...
+        VariableKey v2 = new VariableKey(module, "v2", v2Type, true);
+        RuntimeDiagnosticLog diagnosticLog = new RuntimeDiagnosticLog();
+        ConfigResolver configResolver =
+                new ConfigResolver(Map.ofEntries(Map.entry(module, new VariableKey[]{v1, v2})), diagnosticLog,
+                        List.of(new CliProvider(ROOT_MODULE, "-CmyOrg.test_module.v1=87"),
+                                new TomlFileProvider(ROOT_MODULE,
+                                        getConfigPath("UnsupportedCLITypeConfig.toml"), Set.of(module))));
+        Map<VariableKey, ConfigValue> configValueMap = configResolver.resolveConfigs();
+        Object v1Value = configValueMap.get(v1).getValue();
+        Object v2Value = configValueMap.get(v2).getValue();
```
Comment (HindujaB) [inline bvm/ballerina-runtime/src/test/java/io/ballerina/runtime/test/config/ConfigTest.java:277]: Shall we use `configValueMap.get(v1).getValue()` inside the assertion calls? Creating new variables can be redundant if we don't use them multiple times.

Comment (ravinperera00) [reply bvm/ballerina-runtime/src/test/java/io/ballerina/runtime/test/config/ConfigTest.java:277]: I agree. I'll fix that.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/ConfigUtils.java [hunk 4]
```diff
@@ -85,7 +85,7 @@ public static boolean containsUnsupportedMembers(BUnionType unionType) {
     }
 
     private static boolean isSimpleSequenceType(int tag) {
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/ConfigUtils.java:87]:
```suggestion
    private static boolean isPrimitiveOrSequenceType(int tag) {
```

Comment (ravinperera00) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/ConfigUtils.java]: I missed it. Thank you for pointing that out. Fixed with 2fa8603fa532bca9f27e348eef23ea57ad99a88d

**Review by warunalakshitha — COMMENTED** [review]
LGTM

**Review by HindujaB — APPROVED** [review]
LGTM

**Review by warunalakshitha — APPROVED** [review]

**Pull request merged**